src/UiPath.Workflow.Runtime/Expressions/AssemblyReference.cs

```diff
@@ -14,24 +14,11 @@ namespace System.Activities.Expressions;
 [TypeConverter(TypeConverters.AssemblyReferenceConverter)]
 public class AssemblyReference
 {
-    private const int AssemblyToAssemblyNameCacheInitSize = 100;
-    private const int AssemblyCacheInitialSize = 100;
+    private const int AssemblyToAssemblyNameCacheInitSize = 128;
+    private const int AssemblyCacheInitialSize = 128;
 
-    // cache for Assembly ==> AssemblyName
-    // Assembly.GetName() is very very expensive
-    private static readonly object assemblyToAssemblyNameCacheLock = new();
-
-    // Double-checked locking pattern requires volatile for read/write synchronization
-    private static volatile Hashtable assemblyToAssemblyNameCache;
-
-    // cache for AssemblyName ==> Assembly
-    // For back-compat with VB (which in turn was roughly emulating XamlSchemaContext)
-    // we want to cache a given AssemblyName once it's resolved, so it doesn't get re-resolved
-    // even if a new matching assembly is loaded later.
-
-    // Double-checked locking pattern requires volatile for read/write synchronization
-    private static volatile Hashtable assemblyCache;
-    private static readonly object assemblyCacheLock = new();
+    private static readonly ConcurrentDictionary<Assembly, AssemblyName> assemblyToAssemblyNameCache = new(Environment.ProcessorCount, AssemblyToAssemblyNameCacheInitSize);
+    private static readonly ConcurrentDictionary<AssemblyName, Assembly> assemblyCache = new(Environment.ProcessorCount, AssemblyCacheInitialSize);
```

Comment:
AssemblyName only has reference equality, so different instances having the same semantical value will not match when used as keys. Consider using FullName as key instead.